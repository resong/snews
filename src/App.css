.page {
  margin: 2rem;
}
.interactions {
  text-align: center;
}
/* .banner {
  height: 100vh;
} */
img {
  max-height: 45vh;
}
.title {
  margin: 0.2rem;
}
.subtitle {
  margin: 0.2rem 0 2rem 0;
}
.table {
  margin: 2rem 0;
  border-radius: 0.4rem;
  -moz-border-radius: 0.4rem;
}
.table-header {
  display: flex;
  line-height: 2rem;
  font-size: 1.1rem;
  padding: 0 0.75rem;
  justify-content: space-between;
  font-family: Courier, monotype;
  border: 0.1rem solid #d1d1d1;
  background-color: #d1d1d1;
  font-weight: 500;
}
.table-empty {
  margin: 13rem;
  text-align: center;
  font-size: 1rem;
}
.table-row {
  display: flex;
  line-height: 1.5rem;
  white-space: nowrap;
  padding: 0.75rem;
  background: #ffffff;
  border: 0.08rem solid #e3e3e3;
}
.table-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 0.3rem;
}
.table-header > span:hover {
  background-color: #c0c0c0;
}
.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 0.3rem;
}
.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  cursor: pointer;
  font-family: Courier, monotype;
  font-size: 1.1rem;
  font-weight: 500;
  width: 100%;
}
.button-active {
  font-weight: bold;
}
.button-visible {
  background: #e3e3e3;
}
.button-visible:hover {
  border: 1px solid #cccccc;
  color: #707070;
  background-color: #d2d2d2;
}
.sort-icon {
  margin: 0 0.5rem;
}
.dismiss-btn {
  font-size: 1rem;
  font-weight: normal;
}
.dismiss-btn:hover {
  color: #a3a3a3;
}

@media (max-width: 720px) {
  
  .table-header {
    font-size: 0.8rem;
  }

  .button-inline {
    font-size: 0.8rem;
  }
  
  .dismiss-btn {
    font-size: 0.7rem;
  }
}

@media (max-width: 480px) {
  
  .table-header {
    font-size: 0.7rem;
  }

  .button-inline {
    font-size: 0.7rem;
  }
  
  .dismiss-btn {
    font-size: 0.6rem;
  }
}