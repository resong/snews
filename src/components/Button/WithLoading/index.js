import { withLoading } from "../../../hocs/Loading";
import Button from "../../Button";

export const ButtonWithLoading = withLoading(Button);
