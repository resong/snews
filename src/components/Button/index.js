import React, { Component } from "react";
import PropTypes from "prop-types";

class Button extends Component {
  render() {
    const { onClick, className = "", children } = this.props;

    return (
      <button type="button" onClick={onClick} className={className}>
        {children}
      </button>
    );
  }
}

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired
};

export default Button;
