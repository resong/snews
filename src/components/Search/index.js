import React, { Component } from "react";
import PropTypes from "prop-types";

class Search extends Component {
  componentDidMount() {
    if (this.input) {
      this.input.focus();
    }
  }

  render() {
    const { value, onChange, onSubmit, children } = this.props;

    return (
      <form onSubmit={onSubmit}>
        <input
          type="text"
          value={value}
          onChange={onChange}
          ref={node => {
            this.input = node;
          }}
        />
        <button type="submit" className="button-visible">
          {children}
        </button>
      </form>
    );
  }
}

Search.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired
};

export default Search;
