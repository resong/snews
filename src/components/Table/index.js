import React, { Component } from "react";
import PropTypes from "prop-types";

import Button from "../Button";
import Sort, { SORT_CRITERIA } from "../Sort";

class Table extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sortKey: "NONE",
      isSortReverse: false
    };
  }

  onSort = sortKey => {
    const isSortReverse =
      this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
  };

  render() {
    const { list, onDismiss } = this.props;

    const { sortKey, isSortReverse } = this.state;

    const sortedList = SORT_CRITERIA[sortKey](list);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

    return (
      <div className="table">
        <div className="table-header">
          <span style={{ width: "50%" }}>
            <Sort
              sortKey={"TITLE"}
              onSort={this.onSort}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
            >
              Title
            </Sort>
          </span>
          <span style={{ width: "20%" }}>
            <Sort
              sortKey={"AUTHOR"}
              onSort={this.onSort}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
            >
              Author
            </Sort>
          </span>
          <span style={{ width: "10%" }}>
            <Sort
              sortKey={"COMMENTS"}
              onSort={this.onSort}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
            >
              Comments
            </Sort>
          </span>
          <span style={{ width: "10%" }}>
            <Sort
              sortKey={"POINTS"}
              onSort={this.onSort}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
            >
              Points
            </Sort>
          </span>
          <span style={{ width: "10%" }}>Archive</span>
        </div>

        {reverseSortedList.map(item => (
          <div key={item.objectID} className="table-row">
            <span style={{ width: "50%" }}>
              <a href={item.url}>{item.title}</a>
            </span>
            <span style={{ width: "20%" }}>{item.author}</span>
            <span style={{ width: "10%" }}>{item.num_comments}</span>
            <span style={{ width: "10%" }}>{item.points}</span>
            <span style={{ width: "10%" }}>
              <Button
                className="button-inline dismiss-btn"
                onClick={() => onDismiss(item.objectID)}
              >
                Dismiss
              </Button>
            </span>
          </div>
        ))}
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.num,
      points: PropTypes.num
    })
  ),
  onDismiss: PropTypes.func.isRequired,
  onSort: PropTypes.func,
  sortKey: PropTypes.string,
  isSortReverse: PropTypes.bool
};

export default Table;
