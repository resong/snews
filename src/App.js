import React, { Component } from "react";
import axios from "axios";
import pixelSpeechBubble from "./assets/pixel-speech-bubble.png";
import "./App.css";

import Search from "./components/Search";
import Table from "./components/Table";

import { ButtonWithLoading } from "./containers";

const DEFAULT_QUERY = "redux";
const DEFAULT_HPP = "20";

const PATH_BASE = "https://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";

const updateSearchTopStoriesState = (hits, page) => prevState => {
  const { searchKey, results } = prevState;

  const oldHits = results && results[searchKey] ? results[searchKey].hits : [];

  const updatedHits = [...oldHits, ...hits];

  return {
    results: { ...results, [searchKey]: { hits: updatedHits, page } },
    isLoading: false
  };
};

const updateDismissStoryState = id => prevState => {
  const { searchKey, results } = prevState;
  const { hits, page } = results[searchKey];

  const isNotId = item => item.objectID !== id;

  const updatedHits = hits.filter(isNotId);

  return {
    results: { ...results, [searchKey]: { hits: updatedHits, page } }
  };
};

class App extends Component {
  _isMounted = false;

  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: "",
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false
    };
  }

  needToSearchTopStories = searchTerm => {
    return !this.state.results[searchTerm];
  };

  setSearchTopStories = ({ hits, page }) => {
    this.setState(updateSearchTopStoriesState(hits, page));
  };

  onDismiss = id => {
    this.setState(updateDismissStoryState(id));
  };

  onSearchSubmit = event => {
    this.setState(prevState => {
      const { searchTerm } = prevState;
      return { searchKey: searchTerm };
    });

    const { searchTerm } = this.state;

    if (this.needToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault(); // prevent native browser reload
  };

  onSearchChange = event => {
    this.setState({ searchTerm: event.target.value });
  };

  fetchSearchTopStories = (searchTerm, page = 0) => {
    this.setState({ isLoading: true });

    axios(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }));
  };

  componentDidMount() {
    this._isMounted = true;

    this.setState(prevState => {
      const { searchTerm } = prevState;
      return { searchKey: searchTerm };
    });
    this.fetchSearchTopStories(this.state.searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    const { searchKey, results, searchTerm, error, isLoading } = this.state;

    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;
    const list =
      (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="page">
        <div className="interactions">
          <img src={pixelSpeechBubble} alt="Bits Speech Bubble" />
          <h1 className="title">Snews</h1>
          <h2 className="subtitle">Bits and Bytes from the Internet</h2>

          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
        </div>
        {error ? (
          <section className="interactions">
            <p>Something went wrong!</p>
          </section>
        ) : (
          <Table list={list} onDismiss={this.onDismiss} />
        )}
        <section className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            className="button-visible"
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
          >
            More
          </ButtonWithLoading>
        </section>
        <footer className="interactions">
          <div>
            Pulled from <a href="https://news.ycombinator.com/">Hacker News</a>
          </div>
          <div>2018 &#169; Rebecca Song</div>
        </footer>
      </div>
    );
  }
}

export default App;
