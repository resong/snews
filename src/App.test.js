import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import App from "./App";
import Button from "./components/Button";
import Table from "./components/Table";
import Search from "./components/Search";

Enzyme.configure({ adapter: new Adapter() });

describe("App", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<App />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Search", () => {
  const props = {
    onChange: jest.fn(),
    onSubmit: jest.fn()
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Search {...props}>Search</Search>,

      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Search {...props}>Search</Search>);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Button", () => {
  const props = {
    onClick: jest.fn()
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Button {...props}>Click Me!</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Button {...props}>Click Me!</Button>);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("says click me", () => {
    const element = shallow(<Button {...props}>Click Me!</Button>);

    expect(element.contains("Click Me!")).toBe(true);
  });
});

describe("Table", () => {
  const props = {
    list: [
      { title: "1", author: "1", num_comments: 1, points: 2, objectID: "y" },
      { title: "2", author: "2", num_comments: 1, points: 2, objectID: "z" }
    ],
    sortKey: "TITLE",
    isSortReverse: false,
    onDismiss: jest.fn()
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Table {...props} />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows two items in list", () => {
    const element = shallow(<Table {...props} />);

    expect(element.find(".table-row").length).toBe(2);
  });
});
