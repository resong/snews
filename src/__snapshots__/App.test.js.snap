// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <img
      alt="Bits Speech Bubble"
      src="pixel-speech-bubble.png"
    />
    <h1
      className="title"
    >
      Snews
    </h1>
    <h2
      className="subtitle"
    >
      Bits and Bytes from the Internet
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        className="button-visible"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Title
        </button>
      </span>
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Author
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Comments
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Points
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        Archive
      </span>
    </div>
  </div>
  <section
    className="interactions"
  >
    <div>
      <i
        className="fa fa-spinner fa-spin"
      />
    </div>
  </section>
  <footer
    className="interactions"
  >
    <div>
      Pulled from 
      <a
        href="https://news.ycombinator.com/"
      >
        Hacker News
      </a>
    </div>
    <div>
      2018 Â© Rebecca Song
    </div>
  </footer>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Click Me!
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value={undefined}
  />
  <button
    className="button-visible"
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline dismiss-btn"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline dismiss-btn"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
